Add i:Attached.Icon to MenuItem

sn4k3 opened the pull request:
Implements https://github.com/Projektanker/Icons.Avalonia/issues/16

Note: Wasn't able to make it work with with NativeMenuItem.

SebastianRumohr commented:
Looks very good. Could you please remove the code used for the support of 'NativeMenuItem' as it is not working?

SebastianRumohr reviewed (changes requested):
Looks very good. Could you please remove the code used for the support of 'NativeMenuItem' as it is not working?

sn4k3 commented:
> Looks very good. Could you please remove the code used for the support of 'NativeMenuItem' as it is not working?

Have done

SebastianRumohr commented on demo/MainWindow.axaml:
```diff
@@ -27,55 +27,82 @@
-            <TextBlock Text="fas fa-address-card" Grid.Row="2" />
-            <i:Icon Value="fas fa-address-card" Grid.Row="2" />
-            <Button i:Attached.Icon="fas fa-address-card" Grid.Row="2" />
+    <NativeMenu.Menu>
+        <NativeMenu>
+            <NativeMenuItem Header="Open"/>
+            <NativeMenuItem Header="Recent">
+                <NativeMenuItem.Menu>
+                    <NativeMenu/>
+                </NativeMenuItem.Menu>
+            </NativeMenuItem>
+            <NativeMenuItem Header="Quit"  Gesture="CMD+Q"/>
+        </NativeMenu>
+    </NativeMenu.Menu>
```
Should be removed as well.

sn4k3 replied:
I didn't really test that because i have no mac, but if i recall once i lunch the program it crashed with the attached icon (Even on windows)

SebastianRumohr commented on src/Projektanker.Icons.Avalonia/Icon.axaml.cs:
```diff
@@ -35,6 +36,23 @@ public DrawingImage DrawingImage
             private set => SetAndRaise(DrawingImageProperty, ref _drawingImage, value);
         }
 
+        public IBitmap BitmapImage
+        {
+            get
+            {
+                if (_drawingImage is null) return null;
+                var renderBitmap = new RenderTargetBitmap(new PixelSize((int) _drawingImage.Size.Width, (int) _drawingImage.Size.Height));
+                
+                using (var ctx = renderBitmap.CreateDrawingContext(null))
+                using (var dtx = new DrawingContext(ctx))
+                {
+                    _drawingImage.Drawing.Draw(dtx);
+                }
+
+                return renderBitmap;
+            }
+        }
+
```
Should be removed as well.

sn4k3 replied:
That's just an utility method with no penalty, and was the best way i find to convert the drawing to bitmap.
But it's your call :)

SebastianRumohr reviewed (changes requested):
I was thinking maybe instead of
```xml
<MenuItem
   Command="{Binding ResetToDefaults}"
   Header="Reset to defaults"
   i:Attached.Icon="fas fa-undo-alt">
</MenuItem>
```
we should do it like
```xml
<MenuItem
   Command="{Binding ResetToDefaults}"
   Header="Reset to defaults"
   i:MenuItem.Icon="fas fa-undo-alt">
</MenuItem>
```

I think with `i:MenuItem.Icon` it is easier to understand what is happening. `i:Attached.Icon` sets the content of a control, `i:MenutItem.Icon` sets the icon of a menu item and not its content.

I would merge your PR after the pending removals anyway and split the `Attached.cs` myself.  
What do you think?

sn4k3 commented:
> I think with `i:MenuItem.Icon` it is easier to understand what is happening. `i:Attached.Icon` sets the content of a control, `i:MenutItem.Icon` sets the icon of a menu item and not its content.
> 
> I would merge your PR after the pending removals anyway and split the `Attached.cs` myself. What do you think?

I'm fine with that and makes sense 👍

SebastianRumohr commented:
Will merge the PR so that you will be listed as a contributor.

sn4k3 commented:
> Will merge the PR so that you will be listed as a contributor.

Thanks.

Just a note on readme:

```xml
<Style Selector="MenuItem i|Icon">
    <Setter Property="FontSize" Value="{StaticResource FontSizeNormal}" />
    <Setter Property="Margin" Value="0" />
</Style>
```

That style is not required on a fresh app, better to remove from readme to not lead users to add that to their styles when they don't require to

SebastianRumohr commented:
I tried the demo without the style and it didn't work. The icon was too big. I'll check again.

sn4k3 commented:
> I tried the demo without the style and it didn't work. The icon was too big. I'll check again.

The demo requires because you have global styles on i|Icon which affect menu item too, making them out of bounds.
A new app wont require them and work right away

```xml
<Style Selector="i|Icon">
	<Setter Property="FontSize" Value="24" />
	<Setter Property="Grid.Column" Value="1" />
	<Setter Property="Margin" Value="5" />
</Style>
```

I think that should be bounded to Button like `Button i|Icon` ...

On my app i have created an extended Button that allow me to place a icon and text to ease xmal and get rid of verbose StackPanel.
For anyone who may be usefull here the code:

````C#
using System;
using Avalonia;
using Avalonia.Controls;
using Avalonia.Layout;
using Avalonia.Styling;
using Avalonia.Threading;

namespace UVtools.WPF.Controls;

public class ButtonWithIcon : Button, IStyleable
{
    public enum IconPlacementType : byte
    {
        Left,
        Right,
        Top,
        Bottom
    }

    Type IStyleable.StyleKey => typeof(Button);

    public static readonly StyledProperty<string> TextProperty =
        AvaloniaProperty.Register<ButtonWithIcon, string>(nameof(Text));

    public string Text
    {
        get => GetValue(TextProperty);
        set => SetValue(TextProperty, value);
    }

    public static readonly StyledProperty<IconPlacementType> IconPlacementProperty =
        AvaloniaProperty.Register<ButtonWithIcon, IconPlacementType>(nameof(IconPlacement));

    public IconPlacementType IconPlacement
    {
        get => GetValue(IconPlacementProperty);
        set => SetValue(IconPlacementProperty, value);
    }

    public static readonly StyledProperty<string> IconProperty =
        AvaloniaProperty.Register<ButtonWithIcon, string>(nameof(Icon));

    public string Icon
    {
        get => GetValue(IconProperty);
        set => SetValue(IconProperty, value);
    }

    public static readonly StyledProperty<double> SpacingProperty =
        AvaloniaProperty.Register<ButtonWithIcon, double>(nameof(Spacing), 10);

    public double Spacing
    {
        get => GetValue(SpacingProperty);
        set => SetValue(SpacingProperty, value);
    }

    public ButtonWithIcon()
    {
    }

    protected override void OnInitialized()
    {
        base.OnInitialized();

        Dispatcher.UIThread.Post(() =>
        {
            TextProperty.Changed.Subscribe(_ => RebuildContent());
            IconProperty.Changed.Subscribe(_ => RebuildContent());
            IconPlacementProperty.Changed.Subscribe(_ => RebuildContent());
            RebuildContent();
        }, DispatcherPriority.Loaded);
    }

    public IControl MakeIcon()
    {
        return new Projektanker.Icons.Avalonia.Icon { Value = Icon };
    }

    private void RebuildContent()
    {
        if (string.IsNullOrWhiteSpace(Icon))
        {
            if (!string.IsNullOrWhiteSpace(Text))
            {
                Content = Text;
            }
            return;
        }

        if (string.IsNullOrWhiteSpace(Text))
        {
            if (!string.IsNullOrWhiteSpace(Icon))
            {
                Content = MakeIcon();
            }

            return;
        }

        var panel = new StackPanel
        {
            Spacing = Spacing,
            VerticalAlignment = VerticalAlignment.Stretch,
            Orientation = IconPlacement is IconPlacementType.Left or IconPlacementType.Right 
                ? Orientation.Horizontal
                : Orientation.Vertical
        };

        if(IconPlacement is IconPlacementType.Left or IconPlacementType.Top) panel.Children.Add(MakeIcon());
        panel.Children.Add(new TextBlock{ VerticalAlignment = VerticalAlignment.Center, Text = Text});
        if (IconPlacement is IconPlacementType.Right or IconPlacementType.Bottom) panel.Children.Add(MakeIcon());

        Content = panel;
    }
}
````

SebastianRumohr commented:
> > I tried the demo without the style and it didn't work. The icon was too big. I'll check again.
> 
> The demo requires because you have global styles on i|Icon which affect menu item too, making them out of bounds. A new app wont require them and work right away
> 
> ```
> <Style Selector="i|Icon">
> 	<Setter Property="FontSize" Value="24" />
> 	<Setter Property="Grid.Column" Value="1" />
> 	<Setter Property="Margin" Value="5" />
> </Style>
> ```
> 
> I think that should be bounded to Button like `Button i|Icon` and not left as global and then you can drop the MenuItem styles ...

ah right, Thanks.
 
> On my app i have created an extended Button that allow me to place a icon and text to ease xmal and get rid of verbose StackPanel. 

nice work 👍
